File: stripe/src/main/java/com/stripe/android/view/CountryAdapter.java
package com.stripe.android.view;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.VisibleForTesting;
import android.support.v4.os.ConfigurationCompat;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.Filter;
import android.widget.TextView;

import com.stripe.android.R;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.Locale;

/**
 * Adapter that populates a list of countries for a spinner.
 */
class CountryAdapter extends ArrayAdapter {
    @NonNull private final Context mContext;
    @NonNull private final List<String> mCountries;
    @NonNull private final Filter mFilter;

    private List<String> mSuggestions;

    CountryAdapter(@NonNull Context context, @NonNull List<String> countries) {
        super(context, R.layout.menu_text_view);
        mContext = context;
        mCountries = getOrderedCountries(countries);
        mSuggestions = mCountries;
        mFilter = new Filter() {
            @NonNull
            @Override
            protected FilterResults performFiltering(CharSequence charSequence) {
                final FilterResults filterResults = new FilterResults();
                final List<String> suggestedCountries = new ArrayList<>();
                if (charSequence == null) {
                    filterResults.values = mCountries;
                    return filterResults;
                }
                final String charSequenceLowercase = charSequence.toString()
                        .toLowerCase(Locale.ROOT);
                for (String country : mCountries) {
                    if (country.toLowerCase(Locale.ROOT).startsWith(charSequenceLowercase)) {
                        suggestedCountries.add(country);
                    }
                }
                if (suggestedCountries.size() == 0 || (suggestedCountries.size() == 1 &&
                        suggestedCountries.get(0).equals(charSequence.toString()))) {
                    filterResults.values = mCountries;
                } else {
                    filterResults.values = suggestedCountries;
                }
                return filterResults;
            }

            @SuppressWarnings("unchecked")
            @Override
            protected void publishResults(CharSequence charSequence, FilterResults filterResults) {
                mSuggestions = (List<String>) filterResults.values;
                notifyDataSetChanged();
            }
        };
    }

    @Override
    public int getCount() {
        return mSuggestions.size();
    }

    @Override
    public String getItem(int i) {
        return mSuggestions.get(i);
    }

    @Override
    public long getItemId(int i) {
        return i;
    }

    @NonNull
    @Override
    public View getView(int i, View view, @NonNull ViewGroup viewGroup) {
        if (view instanceof TextView) {
            ((TextView) view).setText(getItem(i));
            return view;
        } else {
            final TextView countryText = (TextView) LayoutInflater.from(mContext).inflate(
                    R.layout.menu_text_view, viewGroup, false);
            countryText.setText(getItem(i));
            return countryText;
        }
    }

    @NonNull
    @Override
    public Filter getFilter() {
        return mFilter;
    }

    @NonNull
    private List<String> getOrderedCountries(@NonNull List<String> countries) {
        // Show user's current locale first, followed by countries alphabetized by display name
        Collections.sort(countries, new Comparator<String>() {
            @Override
            public int compare(String country1, String country2) {
                return country1.toLowerCase(Locale.ROOT)
                        .compareTo(country2.toLowerCase(Locale.ROOT));
            }
        });
        countries.remove(getCurrentLocale().getDisplayCountry());
        countries.add(0, getCurrentLocale().getDisplayCountry());
        return countries;
    }

    @VisibleForTesting
    Locale getCurrentLocale() {
        return ConfigurationCompat.getLocales(mContext.getResources().getConfiguration()).get(0);
    }

}

File: stripe/src/main/java/com/stripe/android/view/CountryAdapter.kt
package com.stripe.android.view

import android.app.Activity
import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.InputMethodManager
import android.widget.ArrayAdapter
import android.widget.Filter
import android.widget.TextView
import com.stripe.android.R
import java.lang.ref.WeakReference
import java.util.Locale

/**
 * Adapter that populates a list of countries for a spinner.
 */
internal class CountryAdapter(
    context: Context,
    initialCountries: List<Country>
) : ArrayAdapter<Country>(context, R.layout.country_text_view) {
    private val countryFilter: Filter = CountryFilter(
        initialCountries,
        this,
        context as? Activity
    )
    private var suggestions: List<Country> = initialCountries

    override fun getCount(): Int {
        return suggestions.size
    }

    override fun getItem(i: Int): Country {
        return suggestions[i]
    }

    override fun getItemId(i: Int): Long {
        return getItem(i).hashCode().toLong()
    }

    override fun getView(i: Int, view: View?, viewGroup: ViewGroup): View {
        return if (view is TextView) {
            view.text = getItem(i).name
            view
        } else {
            val countryText = LayoutInflater.from(context).inflate(
                R.layout.country_text_view, viewGroup, false) as TextView
            countryText.text = getItem(i).name
            countryText
        }
    }

    override fun getFilter(): Filter {
        return countryFilter
    }

    private class CountryFilter(
        private val initialCountries: List<Country>,
        private val adapter: CountryAdapter,
        activity: Activity?
    ) : Filter() {
        private val activityRef = WeakReference(activity)

        override fun performFiltering(constraint: CharSequence?): FilterResults {
            val filterResults = FilterResults()
            filterResults.values = constraint?.let {
                filteredSuggestedCountries(constraint)
            } ?: initialCountries
            return filterResults
        }

        override fun publishResults(
            constraint: CharSequence?,
            filterResults: FilterResults?
        ) {
            val suggestions = filterResults?.values as List<Country>

            activityRef.get()?.let { activity ->
                if (suggestions.any { it.name == constraint }) {
                    hideKeyboard(activity)
                }
            }

            adapter.suggestions = suggestions
            adapter.notifyDataSetChanged()
        }

        private fun filteredSuggestedCountries(constraint: CharSequence?): List<Country> {
            val suggestedCountries = getSuggestedCountries(constraint)

            return if (suggestedCountries.isEmpty() || isMatch(suggestedCountries, constraint)) {
                initialCountries
            } else {
                suggestedCountries
            }
        }

        private fun getSuggestedCountries(constraint: CharSequence?): List<Country> {
            return initialCountries
                .filter {
                    it.name.toLowerCase(Locale.ROOT).startsWith(
                        constraint.toString().toLowerCase(Locale.ROOT)
                    )
                }
        }

        private fun isMatch(countries: List<Country>, constraint: CharSequence?): Boolean {
            return countries.size == 1 && countries[0].name == constraint.toString()
        }

        private fun hideKeyboard(activity: Activity) {
            val inputMethodManager =
                activity.getSystemService(Context.INPUT_METHOD_SERVICE) as? InputMethodManager
            if (inputMethodManager?.isAcceptingText == true) {
                inputMethodManager.hideSoftInputFromWindow(activity.currentFocus?.windowToken, 0)
            }
        }
    }
}
